### DISCUSSION TITLE
How to set keymappings for `cmdline` for nvim-cmp plugin

### DISCUSSION THREAD

**Discussion by dil-mnasir** [LazyVim/LazyVim#2072] [Q&A] [answered]
I have this setup for nvim-cmp, but it seems the custom mappings I defined in `opts.mapping` is not working for cmdline
```
....
    cmp.setup.cmdline(":", {
      mapping = cmp.mapping.preset.cmdline(),
      sources = cmp.config.sources({
        { name = "path" },
        { name = "cmdline" },
        { name = "cmp-cmdline-history" },
      }),
    })
...
```

**Comment by dpetka2001** [comment, marked as answer]
It's not working because it's a different function. For regular `nvim-cmp` the function used is `cmp.mapping.preset.insert` and for `cmdline` is `cmp.mapping.preset.cmdline`. You could just redefine them like so
```lua
        mapping = cmp.mapping.preset.cmdline({
          ["<C-j>"] = {
            c = function(fallback)
              local cmp = require("cmp")
              if cmp.visible() then
                cmp.select_next_item()
              else
                fallback()
              end
            end,
          },
        }),
```
this will use `<C-j>` to select next item in cmdline. See how mapping are defined [here](https://github.com/hrsh7th/nvim-cmp/blob/0b751f6beef40fd47375eaf53d3057e0bfa317e4/lua/cmp/config/mapping.lua#L74) and change them accordingly to the keys you'd like.

**Reply by Eryx5502** [reply]
Just a (silly) question, where should I write this config?

**Reply by dpetka2001** [reply]
Create for example a file in `~/.config/nvim/lua/plugins/cmp.lua` with following contents
```lua
return {
  {
    "hrsh7th/cmp-cmdline",
    config = function()
      local cmp = require("cmp")
      ---@diagnostic disable-next-line: missing-fields
      cmp.setup.cmdline(":", {
        mapping = cmp.mapping.preset.cmdline({
          ["<C-j>"] = {
            c = function(fallback)
              if cmp.visible() then
                cmp.select_next_item()
              else
                fallback()
              end
            end,
          },
        }),
        sources = cmp.config.sources({
          { name = "path" },
        }, {
          {
            name = "cmdline",
            option = {
              ignore_cmds = { "Man", "!" },
            },
          },
        }),
      })
    end,
  },
}
```
Test to see if it's working for you. Most configuration was taken as is from `cmp-cmdline` repo and I just added the mapping part.

**Reply by nipuna-perera** [reply]
How do I get this to work with supertab? I am using the supertab example in LazyVim docs. It's doing a tbl_extend on opts.mapping I am not sure how to add `cmp.mapping.preset.cmdline` to the same.
